### ISSUE TITLE
Implement a dynamic caching layer that establishes watches utilizing a ServiceAccount

### ISSUE THREAD

**Issue by everettraven** [operator-framework/operator-controller#974]
As part of managing installed content and reconciling any changes made to the managed content, we need to configure watches on the resources managed for a given `ClusterExtension`. After #971, given a `ClusterExtension` and the content to be managed, we can configure clients for establishing watches using the `ServiceAccount` specified in the `ClusterExtension`.

While the exact implementation may vary, the currently proposed approach is as follows:

For each ClusterExtension:
- Uses a similar approach as the helm-operator-plugins [client.RestConfigMapper](https://pkg.go.dev/github.com/operator-framework/helm-operator-plugins@v0.2.1/pkg/client#RestConfigMapper) to generate a [rest.Config](https://pkg.go.dev/k8s.io/client-go/rest#Config) that uses a token from the ServiceAccount provided in ClusterExtension.Spec.ServiceAccount.Name
    - In order to support authentication token refreshing during a watch, we will need to do some further customization to the rest.Config resource. In order to achieve this, we will implement a custom [http.RoundTripper](https://pkg.go.dev/net/http#RoundTripper) that will set the Authorization HTTP header to a valid token retrieved from the “TokenGetter” object. This is a similar approach that is taken by [client-go to utilize a BearerTokenFile](https://github.com/kubernetes/client-go/blob/b9309ac26b168965a720c26e3f5329e563ae0c92/transport/round_trippers.go#L304-L318). The custom http.RoundTripper we create could be a lighter weight wrapper around the default used by client-go that can be configured via the rest.Config.WrapTransport option that uses a [transport.WrapperFunc](https://pkg.go.dev/k8s.io/client-go@v0.30.2/transport#WrapperFunc).
- Configures a new controller-runtime [cache.Cache](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/cache#Cache) with the generated [rest.Config](https://pkg.go.dev/k8s.io/client-go/rest#Config)
    - Note: It may be nice to have a lightweight abstraction on top to simplify the logic within the ClusterExtension reconciler for creating new caches and establishing watches for the managed resources
- Uses the [existing approach](https://github.com/operator-framework/operator-controller/blob/208e584ab6be82d11eb0955b0553adbd2244a9fc/internal/controllers/clusterextension_controller.go#L353-L355) for establishing watches with the [controller.Controller](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.4/pkg/controller#Controller) Watch() method
- Starts the cache
- When a ClusterExtension is deleted, it’s cache is stopped and removed

A mermaid graph to visualize the logical flow:
```mermaid
graph TD
    A(ClusterExtension)
    B(Content To Manage)
    C(Managed Content Cache)
    D(RestConfigMapper)
    E(rest.Config)
    F(cache.Cache)
    G(controller.Watch)
    J(ClusterExtension Reconciler)
    L(Watch)

    A -- Provided To --> C
    B -- Provided To --> C
    C -- Fetches Rest Config --> D
    D -- Creates for ServiceAccount --> E
    C -- Creates --> F
    E -- Provided To --> F
    F -- Provided To --> G
    G -- Creates --> L
    L -- Events Trigger --> J
    C -- Establishes Watch With --> G
```

**Acceptance Criteria**:
- A new Go library/type that provided a `ClusterExtension` and Managed Content can establish watches on managed resources using the `ServiceAccount` specified in the `ClusterExtension` spec
    - Will require #971 be completed first. Likely could utilize the implementation of a generic token fetcher from #972 
- Unit tests

**Comment by joelanford** [comment]
Is it possible to wire up and use a controller-runtime cache-delegating `client.Client` per ClusterExtension, where we configure it with an `http.Client` with the custom `http.RoundTripper`?

**Comment by everettraven** [comment]
@joelanford I've updated the RFC and this issue to now reflect the ideal use of the existing controller-runtime caching layer APIs
